ni_measurementlink_service/_internal/parameter/serialization_strategy.py

```diff
@@ -1,118 +1,133 @@
 """Serialization Strategy."""
+from __future__ import annotations
 
-from typing import Any, Callable, Dict
+import sys
+import typing
+from typing import Any, Callable, Dict, cast
 
 from google.protobuf import type_pb2
+from google.protobuf.descriptor import FieldDescriptor
 from google.protobuf.internal import decoder, encoder
-from typing_extensions import TypeAlias
+from google.protobuf.message import Message
 
-InnerEncoder: TypeAlias = Callable[[Callable[[bytes], int], bytes, bool], int]
-InnerDecoder: TypeAlias = Callable[
-    [memoryview, int, int, type_pb2.Field.Kind.ValueType, Dict[int, Any]], int
-]
-EncoderFactory: TypeAlias = Callable[[int], InnerEncoder]
-DecoderFactory: TypeAlias = Callable[[int, str], InnerDecoder]
+if typing.TYPE_CHECKING:
+    if sys.version_info >= (3, 10):
+        from typing import TypeAlias
+    else:
+        from typing_extensions import TypeAlias
 
+WriteFunction: TypeAlias = Callable[[bytes], int]
+Encoder: TypeAlias = Callable[[WriteFunction, bytes, bool], int]
+PartialEncoderConstructor: TypeAlias = Callable[[int], Encoder]
+EncoderConstructor: TypeAlias = Callable[[int, bool, bool], Encoder]
 
-def _scalar_encoder(encoder) -> Callable[[int], InnerEncoder]:
-    """Constructs a scalar encoder factory that takes a field index and returns an InnerEncoder.
+Decoder: TypeAlias = Callable[[memoryview, int, int, Message, Dict[int, Any]], int]
```

Comment:
FYI: it is possible to override a 3rd party type stub by forking it and putting it in the first directory listed in `mypy_path`. We could do this to override decoder.pyi when we run mypy on our project. However, I don't know if running mypy on a measurement service will also need the forked type stub. We shouldn't distribute it as part of our package.